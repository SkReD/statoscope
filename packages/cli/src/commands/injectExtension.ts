import fs from 'fs';
import { Argv } from 'yargs';
import { parseChunked, stringifyStream } from '@discoveryjs/json-ext';
import { waitFinished } from '@statoscope/report-writer/dist/utils';
import type { Webpack } from '@statoscope/webpack-model/webpack';
import { Extension } from '@statoscope/stats/spec/extension';
import { mergeCustomExtensionsIntoCompilation } from '../utils';

export default function (yargs: Argv): Argv {
  return yargs.command(
    'inject-extension',
    `Add an extension into existing stats`,
    (yargs) => {
      return yargs
        .option('input', {
          describe: 'path to a stats.json',
          alias: 'i',
          type: 'string',
        })
        .option('extension', {
          describe:
            'path to json-file with extension (raw json could be passed as stdin)',
          alias: 'e',
          type: 'string',
        })
        .array(['extension'])
        .demandOption('input');
    },
    async (argv) => {
      const extensions: Extension<unknown>[] = [];

      const parsedExtension: Extension<unknown>[] =
        argv.extension?.map((filepath) =>
          JSON.parse(fs.readFileSync(filepath, 'utf-8')),
        ) ?? [];

      if (parsedExtension.length) {
        extensions.push(...parsedExtension.flat());
      } else {
        const stdinReports: Extension<unknown>[] = [
          JSON.parse(fs.readFileSync(0, 'utf-8') || '[]'),
        ];

        extensions.push(...stdinReports.flat());
      }

      const parsed: Webpack.Compilation = await parseChunked(
        fs.createReadStream(argv.input),
      );
      const merged = mergeCustomExtensionsIntoCompilation(parsed, extensions);
      const outputStream = stringifyStream(merged);

      outputStream.pipe(process.stdout);
      await waitFinished(outputStream);
    },
  );
}
