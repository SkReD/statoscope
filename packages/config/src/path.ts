import path from 'path';
import module from 'module';
import chalk from 'chalk';

export enum PackageAliasType {
  PLUGIN = 'PLUGIN',
  REPORTER = 'REPORTER',
}

export function normalizePath(source: string, rootDir: string): string {
  return source.replace('<rootDir>', rootDir);
}

export function makeRequireFromPath(pathname: string): NodeRequire {
  return module.createRequire(path.resolve(pathname, '_'));
}

// package namespace (empty string when not set) followed by the package name (wo namespace)
type PackageName = [string, string];

const packageAliasPrefixes = new Map<PackageAliasType, Record<PrefixType, string>>();

enum PrefixType {
  STATOSCOPE_NS = '@statoscope',
  NON_STATOSCOPE_NS = '@NON_STATOSCOPE_NS',
}

packageAliasPrefixes.set(PackageAliasType.PLUGIN, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-plugin',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-plugin',
});

packageAliasPrefixes.set(PackageAliasType.REPORTER, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-reporter',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-reporter',
});

/**
 * Concatenate namespace (when provided) with the package name.
 */
function getFullPackageName([namespace, name]: PackageName): string {
  if (namespace) {
    return `${namespace}/${name}`;
  }

  return name;
}

/**
 * Return error message based on the list of package aliases (provided by the user and alternative).
 */
function getAliasPackageResolutionError(packageAliases: PackageName[]): string {
  const providedAlias = getFullPackageName(packageAliases[0]);

  let errorMessage = `Can't resolve package ${chalk.yellow.italic(providedAlias)}.\n`;

  const { italic: italicChalk } = chalk;

  if (packageAliases.length > 1) {
    const alternativeAlias = getFullPackageName(packageAliases[1]);

    errorMessage += `Also tried to resolve it with ${italicChalk.yellow(
      alternativeAlias,
    )} alias which didn't work either.\n\n`;
  } else {
    errorMessage += '\n\n';
  }

  errorMessage += 'Try installing the package locally:\n';
  errorMessage += `- with ${italicChalk('npm')}: ${chalk.bgGrey(
    'npm i -D ' + providedAlias,
  )} (or corresponding package alias)\n`;
  errorMessage += `- with ${italicChalk('yarn')}: ${chalk.bgGrey(
    'yarn add -D ' + providedAlias,
  )} (or corresponding package alias)\n`;

  return errorMessage;
}

/**
 * Resolve full or short package name to it's absolute path.
 * @param packageAliasType - Plugin or reporter enum.
 * @param aliasName - Package name (alias). Can be passed in the "full" or "short" form.
 *   Will append prefixes based on packageAliasType value in order to resolve the short form alias.
 * @param fromDir - Directory used as "root" while resolving the package.
 */
export function resolveAliasPackage(
  packageAliasType: PackageAliasType,
  aliasName: string,
  fromDir: string,
): string {
  const localRequire = makeRequireFromPath(fromDir);
  aliasName = normalizePath(aliasName, fromDir);

  if (aliasName.startsWith('.') || path.isAbsolute(aliasName)) {
    localRequire(aliasName);
    return aliasName;
  }

  let packageNamespace = '';
  let packageName = aliasName;

  const packageNameRegex = /^(@.+?)\/(.+)/;

  if (packageNameRegex.test(aliasName)) {
    [, packageNamespace = '', packageName] = aliasName.match(packageNameRegex)!;
  }

  // can have from one up to two entries
  const packageAliases: PackageName[] = [
    [packageNamespace, packageName], // original form
  ];

  const prefixes = packageAliasPrefixes.get(packageAliasType)!;

  const prefix =
    packageNamespace === PrefixType.STATOSCOPE_NS
      ? prefixes[PrefixType.STATOSCOPE_NS]
      : prefixes[PrefixType.NON_STATOSCOPE_NS];

  if (!packageName.startsWith(prefix)) {
    packageAliases.push([packageNamespace, `${prefix}-${packageName}`]);
  }

  const paths = packageAliases.map(([namespace, packageName]) =>
    path.join(namespace, packageName),
  );

  for (const path of paths) {
    try {
      localRequire(path);
      return path;
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  throw new Error(getAliasPackageResolutionError(packageAliases));
}
