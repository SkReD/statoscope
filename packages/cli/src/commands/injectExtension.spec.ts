import fs from 'fs';
import path from 'path';
import os from 'os';
import yargs from 'yargs';
import injectExtension from './injectExtension';

const rootPath = path.resolve(__dirname, '../../../../');
const outputDir = path.join(
  rootPath,
  'test/temp',
  Date.now().toString(),
  path.relative(rootPath, __filename),
);
const statsDir = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/injectReport/stats/',
);
const extDir = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/injectReport/extensions/',
);
const stats = [
  'stats-with-reports.json',
  'stats-without-extensions.json',
  'stats-without-meta.json',
];
const exts = ['multiple-extensions.json', 'single-extension-a.json'];

fs.mkdirSync(outputDir, { recursive: true });

jest.mock('../../package.json', () => ({
  ...jest.requireActual('../../package.json'),
  version: '8.8.8',
}));
jest.mock('@statoscope/stats-extension-custom-reports/package.json', () => ({
  ...jest.requireActual('@statoscope/stats-extension-custom-reports/package.json'),
  version: '8.8.8',
}));

const output: string[] = [];
// eslint-disable-next-line
let consoleLogSpy: any = null;

beforeEach(() => {
  const originalWrite = process.stdout.write.bind(process.stdout);
  consoleLogSpy = jest
    .spyOn(process.stdout, 'write')
    .mockImplementation((data, f, cb) => {
      output.push(
        String(data).replace(process.cwd(), '<pwd>').replace(os.tmpdir(), '<tmp>'),
      );
      return originalWrite(data, f, cb);
    });
});

afterEach(() => {
  consoleLogSpy?.mockRestore();
  output.length = 0;
});

function getOutput(): unknown {
  return JSON.parse(output.join(''));
}

describe.each(stats)('%s,', (filename) => {
  test.each(exts)('%s,', async (ext) => {
    const statsPath = path.join(statsDir, filename);
    const extPath = path.join(extDir, ext);

    let y = yargs(['inject-extension', '--input', statsPath, '-e', extPath]);

    y = injectExtension(y);
    y.fail((_, error) => {
      console.error(error);
    });

    await y.argv;

    expect(getOutput()).toMatchSnapshot();
  });
});
