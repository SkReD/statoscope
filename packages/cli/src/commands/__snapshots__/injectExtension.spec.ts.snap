// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats-with-reports.json, multiple-extensions.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "foo",
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-reports",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension-1",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension-2",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-with-reports.json, single-extension-a.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "foo",
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-reports",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-extensions.json, multiple-extensions.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension-1",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension-2",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-extensions.json, single-extension-a.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-meta.json, multiple-extensions.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension-1",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension-2",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-meta.json, single-extension-a.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "@statoscope/stats-extension-custom-extension",
        },
        "payload": {
          "compilations": [
            {
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;
